packages/gguf/src/gguf.ts

```diff
@@ -0,0 +1,275 @@
+type MetadataBaseValue = string | number | bigint | boolean;
+type MetadataValue = MetadataBaseValue | MetadataBaseValue[] | MetadataValue[]; /// recursive as arrays can be nested.
+
+type Version = 1 | 2 | 3;
+const isVersion = (version: number): version is Version => version === 1 || version === 2 || version === 3;
+
+const ggufMagicNumber = new Uint8Array([0x47, 0x47, 0x55, 0x46]); /// "GGUF"
+
+export enum GGMLQuantizationType {
+	F32 = 0,
+	F16 = 1,
+	Q4_0 = 2,
+	Q4_1 = 3,
+	Q5_0 = 6,
+	Q5_1 = 7,
+	Q8_0 = 8,
+	Q8_1 = 9,
+	Q2_K = 10,
+	Q3_K = 11,
+	Q4_K = 12,
+	Q5_K = 13,
+	Q6_K = 14,
+	Q8_K = 15,
+	IQ2_XXS = 16,
+	IQ2_XS = 17,
+	IQ3_XXS = 18,
+	IQ1_S = 19,
+	IQ4_NL = 20,
+	IQ3_S = 21,
+	IQ2_S = 22,
+	IQ4_XS = 23,
+}
+
+enum GGUFValueType {
+	UINT8 = 0,
+	INT8 = 1,
+	UINT16 = 2,
+	INT16 = 3,
+	UINT32 = 4,
+	INT32 = 5,
+	FLOAT32 = 6,
+	BOOL = 7,
+	STRING = 8,
+	ARRAY = 9,
+	UINT64 = 10,
+	INT64 = 11,
+	FLOAT64 = 12,
+}
+function isGGUFValueType(n: number): n is GGUFValueType {
+	return typeof GGUFValueType[n] === "string";
+}
+
+const HTTP_CHUNK_SIZE = 2 * 10 ** 6; /// 2MB
+const HTTP_DATA_LEEWAY = 5 * 10 ** 5; /// 500kb
+
+/**
+ * Internal stateful instance to fetch ranges of HTTP data when needed
+ */
+class RangeView {
+	private chunk: number;
+	private buffer: ArrayBuffer;
+
+	readonly view: DataView;
+
+	constructor(public url: string) {
+		this.chunk = 0;
+		/// TODO(fix typing)
+		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+		// @ts-ignore
+		this.buffer = new ArrayBuffer(0, { maxByteLength: 50 * 10 ** 6 });
+		this.view = new DataView(this.buffer);
+	}
+	/**
+	 * Fetch a new chunk from the server
+	 */
+	async fetchChunk() {
+		const range = [this.chunk * HTTP_CHUNK_SIZE, (this.chunk + 1) * HTTP_CHUNK_SIZE - 1];
+		const buf = new Uint8Array(
+			await (
+				await fetch(this.url, {
+					headers: {
+						Range: `bytes=${range[0]}-${range[1]}`,
+					},
+				})
+			).arrayBuffer()
+		);
+		/// TODO(fix typing)
+		// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+		// @ts-ignore
+		this.buffer.resize((this.chunk + 1) * HTTP_CHUNK_SIZE);
+		new Uint8Array(this.buffer).set(buf, this.chunk * HTTP_CHUNK_SIZE);
+		this.chunk += 1;
+	}
+	/**
+	 * Check whether we need to fetch a new chunk
+	 */
+	async check(offset: number) {
+		if (this.view.byteLength - offset < HTTP_DATA_LEEWAY) {
+			await this.fetchChunk();
+		}
+	}
+}
+
+function readVersionedSize(view: DataView, byteOffset: number, version: Version): bigint {
+	switch (version) {
+		case 1: {
+			const n = view.getUint32(byteOffset, true);
+			return BigInt(n);
+		}
+		case 2:
+		case 3: {
+			return view.getBigUint64(byteOffset, true);
+		}
+	}
+}
+
+function readString(view: DataView, offset: number): { value: string; newOffset: number } {
+	const length = view.getBigUint64(offset, true);
+	const value = new TextDecoder().decode(view.buffer.slice(offset + 8, offset + 8 + Number(length)));
+	return { value, newOffset: offset + 8 + Number(length) };
```

Comment:
instead of returning `{value, newOffset}`, imo it should return `{value, length}`
and then instead of using below `offset = keyResult.newOffset;` we can use `offset += keyResult.length;`

same applied to `readMetadataValue` below

Reply:
`length` is better than `newOffset` I agree. I will be making that change to `hyllama` too.

Reply:
I tried doing it last friday but then it was failing my E2Es. 

I'll try again tomorrow

Reply:
https://github.com/huggingface/huggingface.js/pull/543